### PR TITLE
Implement invocation retries for VM and other failures

### PR DESCRIPTION
**Pull request by alevy**
This PR adds invocation retries in the worker. Each invocation is retried up to 5 times if there is any error in the invocation, before returning either a successful result or an error.

### CODE REVIEW

File: snapfaas/src/vm.rs [hunk 1]
```diff
@@ -386,18 +386,32 @@ impl Vm {
...
+                    let timestamps = RequestTimestamps {
+                        at_vmm: precise_time_ns(),
+                        request: req.clone(),
+                        ..Default::default()
+                    };
+
+                    let (tx, rx) = mpsc::channel();
+                    if invoke_handle.send(Message::Request((req.clone(), tx, timestamps))).is_err() {
+                        return;
+                    }
```
Comment (tan-yue) [inline snapfaas/src/vm.rs:406]: I think this SendError is fatal. `send` errors only when the resource manager side receiver is disconnected.
Plus, we are panicking in multivm/main.rs

**Comment by tan-yue** [comment]
I don't think the two places doing retries now are the best option. The system has such a architecture that every request is handled by a worker in the workerpool: both external and internal (invoke) requests are sent to the `mpsc::channel` all workers listen on. Thus, I think worker is the better place to do retries.

Another thing which is more like a memo:
I intended for the return value of `send_req` to tell whether the request is successfully put into the `mpsc::channel`. Now I think about it, it doesn't do much and we probably should panic when `send` errors instead of returning false. `send` erring is fatal because it means workers cannot read from the `mpsc::channel`.

**Comment by alevy** [comment]
@tan-yue revised to be in `worker.rs` instead of two separate places. It's indeed cleaner and easier to understand this was, I think.

File: snapfaas/src/worker.rs [hunk 2]
```diff
@@ -27,6 +27,93 @@ pub struct Worker {
...
+                let id = thread::current().id();
+                let status = match e {
+                    resource_manager::Error::InsufficientEvict |
+                    resource_manager::Error::LowMemory(_) => {
+                        error!("[Worker {:?}] Resource exhaustion", id);
+                        RequestStatus::ResourceExhausted
+                    }
+                    resource_manager::Error::FunctionNotExist=> {
+                        error!("[Worker {:?}] Requested function doesn't exist: {:?}", id, function_name);
+                        RequestStatus::FunctionNotExist
```
Comment (tan-yue) [inline snapfaas/src/worker.rs:95]: A minor optimization: we can break out of loop if the error is FunctionNotExist. Retrying cannot solve it.

**Pull request merged**